sonarts-core/src/rules/consecutiveOverloadsRule.ts

```diff
@@ -0,0 +1,154 @@
+import * as ts from "typescript";
+import { SonarRuleMetaData } from "../sonarRule";
+import { TypedSonarRuleVisitor } from "../utils/sonarAnalysis";
+import { is, isLiteralExpression, isSignatureDeclaration } from "../utils/nodes";
+
+export class Rule extends tslint.Rules.TypedRule {
+  public static metadata: SonarRuleMetaData = {
+    ruleName: "consecutive-overloads",
+    description: "Method overloads should be grouped together",
+    rationale: tslint.Utils.dedent``,
+    optionsDescription: "",
+    options: null,
+    rspecKey: "RSPEC-4136",
+    type: "functionality",
+    typescriptOnly: false,
+  };
+
+  public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): tslint.RuleFailure[] {
+    return new Visitor(this.getOptions().ruleName, program).visit(sourceFile).getIssues();
+  }
+}
+
+class Visitor extends TypedSonarRuleVisitor {
+  protected visitTypeLiteral(node: ts.TypeLiteralNode): void {
+    this.checkOverloads(node.members);
+    super.visitTypeLiteral(node);
+  }
+
+  protected visitInterfaceDeclaration(node: ts.InterfaceDeclaration): void {
+    this.checkOverloads(node.members);
+    super.visitInterfaceDeclaration(node);
+  }
+
+  protected visitClassDeclaration(node: ts.ClassDeclaration): void {
+    this.checkOverloads(node.members);
+    super.visitClassDeclaration(node);
+  }
+
+  protected visitModuleDeclaration(node: ts.ModuleDeclaration): void {
+    const moduleBody = node.body;
+    if (is(moduleBody, ts.SyntaxKind.ModuleBlock)) {
+      const functionLike = (moduleBody as ts.ModuleBlock).statements.filter(stmt => isSignatureDeclaration(stmt));
+      this.checkOverloads(functionLike);
+    }
+    super.visitModuleDeclaration(node);
+  }
+
+  protected visitSourceFile(node: ts.SourceFile): void {
+    this.checkOverloads(node.statements.filter(stmt => isSignatureDeclaration(stmt)));
+    super.visitSourceFile(node);
+  }
+
+  private checkOverloads(nodes: ReadonlyArray<ts.Node>) {
+    const misplacedOverloads = getMisplacedOverloads(nodes, declaration =>
+      getOverloadKey(declaration as ts.SignatureDeclaration),
+    );
+    misplacedOverloads.forEach(overload =>
+      this.addIssue(overload, `All '${printOverload(overload)}' signatures should be adjacent`),
+    );
+  }
+}
+
+/** 'getOverloadName' may return undefined for nodes that cannot be overloads, e.g. a `const` declaration. */
+function getMisplacedOverloads(
+  overloads: ReadonlyArray<ts.Node>,
+  getKey: (node: ts.Node) => string | undefined,
+): ts.SignatureDeclaration[] {
+  const result: ts.SignatureDeclaration[] = [];
+  let lastKey: string | undefined;
+  const seen = new Set<string>();
+  for (const node of overloads) {
+    if (node.kind === ts.SyntaxKind.SemicolonClassElement || isAccessor(node)) {
+      continue;
+    }
+
+    const key = getKey(node);
+    if (key !== undefined) {
+      if (seen.has(key) && lastKey !== key) {
+        result.push((node as any) as ts.SignatureDeclaration);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Type declaration of 'any' loses type-safety. Consider replacing it with a more precise type. ([typescript:S4204](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS4204))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&pullRequest=643&issues=AWNpQliMU2gg8QXllzFQ&open=AWNpQliMU2gg8QXllzFQ)

[ISSUE_KEY]: AWNpQliMU2gg8QXllzFQ